### PR TITLE
Fix DockerFile GPU CUDA GPG Keys

### PR DESCRIPTION
**Pull request by Arka161**
The Docker in the repository no longer works, as NVIDIA updated their GPG keys in [Apr 2022](https://developer.nvidia.com/blog/updating-the-cuda-linux-gpg-repository-key/). 

The dockerfile has been updated here

### CODE REVIEW

**Comment by Arka161** [comment]
Hello @jameslamb ,

Thank you for your comment on the PR. I understand that `master` -> `master` PRs are not encouraged for this Repo. Would you like me to close this PR, and then make a new one with a different branch pointing the LightGBM/master? 

Also, I would like to highlight that the current `Docker` image does not work, hence my changes to the GPG key. Let me know if you would like me to change anything :)

**Comment by jameslamb** [comment]
> then make a new one with a different branch

nope, it's ok! You can keep this one. But once this is merged, I recommend deleting and re-creating your fork. (or using `git reset` + `git push --force` to rewrite history on `master`  of your fork)

**Comment by Arka161** [comment]
Thank you for your reply, I understand what you meant! I did not have any other code modified in the origin of my LightGBM fork, which is why I had made it from master. In the future, I will make the PR from a branch created out of master/main :) 

For testing Docker images: One quick idea is trying it on Google Colab / Free GPU instances on Docker, I have tested it on my environment with an RTX 2060 + official NVIDIA drivers (Ubuntu), and it seems to work great, I cannot think of any way to write tests for docker images.

**Comment by Arka161** [comment]
I personally don't like this line the Dockerfile I had to add (looks too hardcoded): 

```
RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys A4B469963BF863CC
```

Though it errors out without it on any system due to NVIDIA deprecating its keys,

**Comment by StrikerRUS** [comment]
> Edit: I have removed the hardcoded condition to syntax NVIDIA recommends on their blog post

Here it is: https://developer.nvidia.com/blog/updating-the-cuda-linux-gpg-repository-key/.

File: docker/gpu/dockerfile.gpu [hunk 1]
```diff
@@ -73,8 +77,8 @@ RUN echo "export PATH=$CONDA_DIR/bin:"'$PATH' > /etc/profile.d/conda.sh && \
     /bin/bash ~/miniforge.sh -b -p $CONDA_DIR && \
     rm ~/miniforge.sh
 
-RUN conda config --set always_yes yes --set changeps1 no && \
-    conda create -y -q -n py3 numpy scipy scikit-learn jupyter notebook ipython pandas matplotlib
```
Comment (StrikerRUS) [inline docker/gpu/dockerfile.gpu:77]: Could you please clarify why these changes were needed?

**Review by StrikerRUS — APPROVED** [review]

File: docker/gpu/dockerfile.gpu (hunk 1 above)
Comment (Arka161) [reply docker/gpu/dockerfile.gpu:77]: Hello, 

I personally thought it would be good to add this because if any error is thrown, you previously could not see what is going on, or where the script is stuck (as it was hidden), so I have made the logs during conda runs visible :) 

Thank you for reviewing the merge request, and let me know if I should change this line back to the original line.

Comment (StrikerRUS) [reply docker/gpu/dockerfile.gpu:77]: Thanks for the clarification!

Please revert changes in lines 76-77 because they are unrelated to the issue this PR is solving.

If you think that it's better to make conda installation more verbose, please open **a separate PR** and make corresponding changes in all Docker files in this repo:
https://github.com/microsoft/LightGBM/blob/83627ff0d6baf49e3b18991f47b12d9fa5724cbc/docker/dockerfile-python#L19-L21
https://github.com/microsoft/LightGBM/blob/83627ff0d6baf49e3b18991f47b12d9fa5724cbc/docker/gpu/dockerfile.gpu#L76-L77

**Review by jameslamb — CHANGES_REQUESTED** [review]
@Arka161 can you please implement @StrikerRUS 's request in https://github.com/microsoft/LightGBM/pull/5369/files#r933896510?

(removing unrelated changes about log verbosity and shell output)

File: docker/gpu/dockerfile.gpu (hunk 1 above)
Comment (Arka161) [reply docker/gpu/dockerfile.gpu]: I am so sorry for the delay, the change has been made, opening the other Pull Request now :) @StrikerRUS

**Comment by Arka161** [comment]
@jameslamb I have made the changes, I am very sorry for the delay as I was travelling. Making the other pull request too.

**Comment by jameslamb** [comment]
No problem! Thanks very much.

I started testing this on a `g4dn` instance on AWS last night. Didn't quite finish, will try in the next few days.

@Arka161 could you clarify what **specifically** you mean by "The Docker in the repository no longer works"? When I was testing on an AWS EC2 instance last night, I was able to build an image from `docker/gpu/dockerfile.gpu` on latest `master`, without any modifications.

**Comment by Arka161** [comment]
Hello @jameslamb : I get [this kind of an error](https://askubuntu.com/questions/13065/how-do-i-fix-the-gpg-error-no-pubkey) when I run the GPU Dockerfile on the master branch of LightGBM. 

Basically, something like: 

```
W: GPG error: ___ trusty InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY ___
```

I got this on my server and my personal Ubuntu machine, and I was unable to build the image. It appeared that NVIDIA changed their CUDA Linux GPG keys, which was the fix for the error.

**Comment by jameslamb** [comment]
Ok thanks for that. I saw that exact same *warning* (that is what the `W: ` means), but was still able to build the image on Amazon Linux.

Are you working in an environment where anything writing to stderr causes a process to exit with an error? Like on Windows using Powershell?

Anyway, I'll come back to this some time in the next few days, once I've had a chance to finish testing, with a reproducible example and a more thorough review. Thanks for your patience.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Sorry for the very long delay! I wanted to be sure to test this, and it took me a while to set up a test environment. We appreciate your patience.

Now that I've done that, I'm sorry to say that unless you provide some information that reveals that I made a mistake in testing,  **I don't support the changes in this PR**.

@Arka161 how exactly did you test these changes? Please be as specific as possible.

It's strange to me that image-building is failing for you on `master` (https://github.com/microsoft/LightGBM/pull/5369#issuecomment-1205738517) and succeeding on this PR, when I see the opposite. It's also surprising to me that the changes proposed in this PR are working for you.

* I'd expect that any patching of keys needs to happen *before* `apt-get update` and `apt-get install`, but you've placed it after those commands.
* with only the changes in this PR, I don't understand how the use of `conda activate` could present an error for me but not for you

Maybe the difference between how you tested this change and how I tested it (described below) will reveal what I'm missing.

### How I tested

I got a GPU-enabled EC2 instance on AWS with the following specifications.

- **AMI**: `Ubuntu 18.04 (ami-0b152cfd354c4c7a4)`
- **instance type**: `g4dn.xlarge` (1 NVIDIA T4 GPU)
- **region**: `us-west-2`
- **root volume size**: 100GB

Set up docker and CUDA drivers on it.

<details><summary>CUDA setup (click me)</summary>

```shell
# following https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html

# install docker
sudo apt-get update -y
sudo apt-get install -y \
    ca-certificates \
    curl \
    gnupg \
    lsb-release

curl -fsSL https://download.docker.com/linux/ubuntu/gpg \
| sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg

echo \
  "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu \
  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null

sudo apt-get update -y
sudo apt-get install -y \
    docker-ce \
    docker-ce-cli \
    containerd.io

sudo systemctl enable docker
sudo systemctl start docker

# install nvidia-docker
distribution=$(. /etc/os-release;echo $ID$VERSION_ID) \
      && curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
      && curl -s -L https://nvidia.github.io/libnvidia-container/$distribution/libnvidia-container.list | \
            sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
            sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list

distribution=$(. /etc/os-release;echo $ID$VERSION_ID) \
    && curl -s -L https://nvidia.github.io/libnvidia-container/gpgkey | sudo apt-key add - \
    && curl -s -L https://nvidia.github.io/libnvidia-container/$distribution/libnvidia-container.list | sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list

distribution=$(. /etc/os-release;echo $ID$VERSION_ID) \
    && curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey | sudo apt-key add - \
    && curl -s -L https://nvidia.github.io/nvidia-docker/$distribution/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list

# get nvidia drivers and tools
# (versions selected by choosing the newest thing shown in `apt search nvidia-driver`)
sudo add-apt-repository ppa:graphics-drivers/ppa --yes
sudo apt-get update -y
sudo apt-get install -y \
    nvidia-driver-515 \
    nvidia-utils-515

# reboot the machine
sudo reboot
```

</details>

Then, set up `nvidia-docker`.

<details><summary>nvidia-docker setup (click me)</summary>

```shell
# back in, in a new session, check that drivers were set up correctly
nvidia-smi

# install nvidia-docker
sudo apt-get update -y
sudo systemctl stop docker
sudo apt-get install -y \
    nvidia-docker2
sudo systemctl restart docker

# check that nvidia-docker was set up correctly,
# and works for the base image relevant to this PR
sudo docker run \
    --rm \
    --gpus all \
    nvidia/cuda:8.0-cudnn5-devel \
    nvidia-smi
```

</details>

Next, created a small Python test script that trains a binary classification model using LightGBM's GPU support.

<details><summary>test-training.py (click me)</summary>

```shell
mkdir -p "${HOME}/repos/pr-5369"
cd "${HOME}/repos/pr-5369"

cat << EOF > ./test-training.py
import lightgbm as lgb
from sklearn.datasets import load_breast_cancer

X, y = load_breast_cancer(return_X_y=True)

dtrain = lgb.Dataset(
    data=X,
    label=y
)

model = lgb.train(
    params={
        "objective": "binary",
        "num_iterations": 10,
        "verbose": 1,
        "num_leaves": 8,
        "device": "gpu"
    },
    train_set=dtrain,
)
print("--- done training ---")
EOF
```

</details>

Next, I downloaded two versions of `dockerfile.gpu`...the latest one on `master` and the one from this pull request.

```shell
# Dockerfile from master
LGBM_COMMIT=c7102e56b246cc5cd73d9787b2c837c0bc384d1e
curl \
    https://raw.githubusercontent.com/microsoft/LightGBM/${LGBM_COMMIT}/docker/gpu/dockerfile.gpu \
> ./Dockerfile-upstream

# Dockerfile from this PR
curl \
    https://raw.githubusercontent.com/Arka161/LightGBM/master/docker/gpu/dockerfile.gpu \
> ./Dockerfile-pr
```

I built an image using the one on `master`, then ran the test script in it.

<details><summary>testing dockerfile.gpu from upstream (click me)</summary>

```shell
sudo docker build \
    -f ./Dockerfile-upstream \
    -t lightgbm-gpu:upstream \
    .

sudo docker run \
    --gpus all \
    --rm \
    --entrypoint="" \
    -v $(pwd):/LightGBM \
    --workdir /LightGBM \
    -it lightgbm-gpu:upstream \
    /bin/bash -c "source activate py3 && python test-training.py"
```

</details>

Here's what I observed.

1. image built successfully, with two warnings relevant to this PR:

> **Warning 1: GPG key**
>
> W: GPG error: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64  InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
> W: The repository 'https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64  InRelease' is not signed.
>
> **Warning 2: source activate deprecated**
>
> DeprecationWarning: 'source deactivate' is deprecated. Use 'conda deactivate'.

2. training ran successfully, and used the GPU. Logs:

```text
/opt/miniforge/envs/py3/lib/python3.10/site-packages/lightgbm/engine.py:177: UserWarning: Found `num_iterations` in params. Will use it instead of argument
  _log_warning(f"Found `{alias}` in params. Will use it instead of argument")
[LightGBM] [Info] Number of positive: 357, number of negative: 212
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 5676
[LightGBM] [Info] Number of data points in the train set: 569, number of used features: 30
[LightGBM] [Info] Using GPU Device: Tesla T4, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 30 dense feature groups (0.02 MB) transferred to GPU in 0.001001 secs. 0 sparse feature groups
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.627417 -> initscore=0.521150
[LightGBM] [Info] Start training from score 0.521150
--- done training ---
```

Then, I repeated this exercise using the Dockerfile from this PR.

<details><summary>testing dockerfile.gpu from this PR (click me)</summary>

```shell
sudo docker build \
    -f ./Dockerfile-pr \
    -t lightgbm-gpu:pr \
    .

sudo docker run \
    --gpus all \
    --rm \
    --entrypoint="" \
    -v $(pwd):/LightGBM \
    --workdir /LightGBM \
    -it lightgbm-gpu:pr \
    /bin/bash -c "source activate py3 && python test.py"
```

</details>

Here's what I observed.

1. The exact same warning about the NVIDIA GPG key is still raised.

> W: GPG error: https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64  InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY A4B469963BF863CC
> W: The repository 'https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64  InRelease' is not signed.

2. A NEW warning is raised by the newly-added code.

> A deprecated public CUDA GPG key appears to be installed.
> To remove the key, run this command:
> sudo apt-key del 7fa2af80

3. The warning about `source` vs. `conda` is still raised.

> DeprecationWarning: 'source deactivate' is deprecated. Use 'conda deactivate'.

4. **Building the image fails**, with the following error.

> E: Conflicting values set for option Signed-By regarding source https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1604/x86_64/ /
> E: The list of sources could not be read.

If I revert all the `apt` related changes and just make the `source` to `conda` changes, **building this image fails**.

> CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
> To initialize your shell, run...

**Comment by jameslamb** [comment]
I'm going to close this PR based on https://github.com/microsoft/LightGBM/pull/5369#pullrequestreview-1064356007 and due to lack of response.

@Arka161 if you see an issue with the conclusions in that comment please do leave a comment here and we can re-open this pull request.

Thanks for your interest in LightGBM, come back and contribute any time!
